Carousel Components and its BaseItemsControl generic class

HClausing opened the pull request:
NOTE: It's the same from #931 with an updated dev branch,

UPDATED:
 - Added **slide effect**;
 - Added **SwiptArea** component for page changing with gestures (thanks to @tungi52 ).

Here's the 99999999 attempt on Carousel component development. 

There's also tow abstract (and generic) classes that can be used on anothers components with a similar behavior, like Stepper, Wizard, and so on. The first one controls the child components (ex: CarouselItem) list, Selected Index, and navigation methods, like Next, Previous, MoveTo, etc. The second one adds DataBinding support with an IEnumerable<TData> parameter.

This Carousel components uses the DataBinding version of the abstract classes, so it can be used on theses ways:

```
<Carousel TData="object">
    <CarouselItem>
            ...
    </CarouselItem>
    <CarouselItem>
    </CarouselItem>
</Carousel>
```

Or:

```
<Carousel ItemsSource="@YourDataSource">
    <ItemTemplate>
            ...
    </ItemTemplate>
</Carousel>
```

![carousel](https://user-images.githubusercontent.com/15158923/108398260-95432b00-71f7-11eb-8203-a6ea65413e32.gif)

Any review are appreciated.

**On Slide effect, there's a throuble: Even the main div clipping the inner component when it's sliding, we can see a temporary horizontal scrollbar until transition ends. Any help are appreciated here.**  :

![slide_scroll_bug](https://user-images.githubusercontent.com/15158923/113571966-b5834980-95ed-11eb-8e1f-c1581e7d06c8.gif)

HClausing committed:
- Add: BaseItemsControl for use with any king of items host component;
 - Add: Carousel component with slide and fade transitions;
 - Add: Docs Samples for Carousel component.

porkopek commented:
> On Slide effect, there's a throuble: Even the main div clipping the inner component when it's sliding, we can see a temporary horizontal scrollbar until transition ends. Any help are appreciated here. :

I didn't see the code, but maybe it's something solvable with `overflow:hidden` ?

HClausing commented:
> > On Slide effect, there's a throuble: Even the main div clipping the inner component when it's sliding, we can see a temporary horizontal scrollbar until transition ends. Any help are appreciated here. :
> 
> I didn't see the code, but maybe it's something solvable with `overflow:hidden` ?

Well, maybe. But the scrollbar that I mean is on content host, and not in the Carousel component, so if the page or any other parent needs to scroll horizontally, It can be a problem.

unchase commented on src/MudBlazor/Base/MudBaseItemsControl.cs:
```diff
@@ -0,0 +1,129 @@
+
+    {
+        /// <summary>
+        /// Colletion of T
```
"Colletion" > "Collection"

unchase commented on src/MudBlazor/Base/MudBaseItemsControl.cs:
```diff
@@ -0,0 +1,129 @@
+
+    {
+        /// <summary>
+        /// Colletion of T
```
Colletion -> Collection

HClausing replied:
Thanks!

HClausing committed:
Fixed wrong typing on MudBaseItemsControl's attribute documentation.

porkopek commented:
With 
```css
.mud-carousel {
    overflow: hidden;
}
```
you fix the overflow.
Other than that, could you do the default transition continuos? I mean, when the second slide entries, the first moves to the left until disappear
![2](https://user-images.githubusercontent.com/13745954/113930846-b5e83400-97e9-11eb-887d-e0f9eaf6269e.gif)

I like the enum `Transition` to specify the animation/transition

porkopek commented on src/MudBlazor/Base/MudBaseItemsControl.cs:
```diff
@@ -0,0 +1,129 @@
+            return base.OnAfterRenderAsync(firstRender);
+        }
+
+        internal bool move_next = true;
```
I think we are using `_moveNext` or `moveNext` as naming convention, you also use it along your code, without the middle `_`. Maybe @mikes-gh  can confirm

mikes-gh replied:
Yes _moveNext

HClausing replied:
Got it. It's according to convention now.

porkopek commented on src/MudBlazor/Components/Carousel/MudCarousel.razor.cs:
```diff
@@ -0,0 +1,160 @@
+        /// Gets or Sets the Template for the Right Arrow
+        /// </summary>
+        [Parameter]
+        public RenderFragment RightArrowTemplate { get; set; }
```
I like everything is a template

HClausing replied:
Thanks! It give us more styling freedom, and I'm sure that someone would request it on the future.

porkopek commented on src/MudBlazor/Styles/components/_carousel.scss:
```diff
@@ -0,0 +1,67 @@
+    position: relative;
+    margin: 0px !important;
+    clip-path: inset(0px 0px 0px 0px);
+}
```
You need `overflow: hidden;` here

HClausing replied:
Nice! worked like a charm! Thanks!

HClausing commented:
> With
> 
> ```css
> .mud-carousel {
>     overflow: hidden;
> }
> ```
> 
> you fix the overflow.
> Other than that, could you do the default transition continuos? I mean, when the second slide entries, the first moves to the left until disappear
> ![2](https://user-images.githubusercontent.com/13745954/113930846-b5e83400-97e9-11eb-887d-e0f9eaf6269e.gif)
> 
> I like the enum `Transition` to specify the animation/transition

Good idea, I'll try this add an instance of RenderFragment to keep "old item" in the screen before item selection changed. then it will start the animation at the same time.

HClausing committed:
- New transition system with exiting animation;
- Custom transitions.

HClausing committed:
Fix comments on <style> section of Custom Transition Example

HClausing commented:
Updates

_- New transition system with exiting animation!_

![new-transition-mode](https://user-images.githubusercontent.com/15158923/114067971-c644f100-9873-11eb-888c-b68ae40e8439.gif)

_- Custom transitions!_

![custom-transition](https://user-images.githubusercontent.com/15158923/114067996-cd6bff00-9873-11eb-8eaa-701b0e48b758.gif)

porkopek commented on src/MudBlazor.Docs/Pages/Components/Carousel/Examples/CarouselCustomTransitionExample.razor:
```diff
@@ -0,0 +1,104 @@
+<styles>
+
+    @*KeyFrames are commented because they can't be declared into Blazor Components.
+    Please paste then into your CSS file.*@
```
You can. Just double the @, that it's the way the Razor engine escapes the `@`

 so it's `@@keyframes`

porkopek replied:
It would be nice if you could use https://animate.style/
```
<link
    rel="stylesheet"
    href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/4.1.1/animate.min.css"
  />
``` 
in the examples, just to show custom animations, but maybe it's not possible, I don't know

HClausing replied:
Yeah, I used the boundIn and boundOut from there!

I'll adjust the style tag on razor component. Thanks.

HClausing replied:
Well, the @@ didn't work:

![image](https://user-images.githubusercontent.com/15158923/114080931-51c57e80-9882-11eb-85ee-327f3cfa5946.png)

HClausing replied:
Oh sorry, I was typing <styles> ... its <style> (no plural)

HClausing replied:
Done, added link for animate.min.css on razor file. Now every animation class uses keyframes from there.

Added a new example with transition settings per page.

Thanks!

HClausing committed:
- Added animation keyframes from https://animate.style/ ;
 - Removed obsolete _carousel.scss, with contend set on <style/> tag of every example;
 - Removed @import reference from _carousel.scss;
 - Added per page transition value example.

henon reviewed (changes requested):
The code looks quite solid. We could even reuse the MudBaseItemsControl as parent of some other components like Chipset, etc. 

The only thing that is missing in this PR are some tests. You know, to make sure nobody will break this beautiful component in future PRs.

HClausing commented on src/MudBlazor.Docs/Pages/Components/Carousel/Examples/CarouselBindingExample.razor:
```diff
@@ -0,0 +1,22 @@
+
+<MudCarousel ItemsSource="@source" Style="height:200px;" ShowArrows="@arrows" ShowDelimiters="@delimiters" AutoCycle="@autocycle">
+    <ItemTemplate>
+        <div class="d-flex flex-column d-flex flex-column justify-center" style="height:100%">
```
Oh, I'm sorry. I fixed it and will be available on next commit.

HClausing commented:
Thanks for the review @henon . I'll write some tests. I always forget about them!

By the way, Do you have some suggestion of what test are important to run in these classes? I'm thinking on check the Move actions, cycle timer, etc.

henon commented:
The best way is to design the tests so that they run through most of the logic. @mikes-gh and I are also using test coverage tools (i.e. DotCover, I forgot whichone Mike uses) to find which parts are left uncovered by tests.  

I'd make sure the methods of MudBaseItemsControl are checked (like SelectedItem, SelectedIndex, MoveTo, Next, Previous, etc). It is also generally good to re-enforce whatever the docs show in their examples with tests so that they won't break unnoticed. Testing the slide timing is probably a bit difficult to get it working reliably on the build server so I am not sure about that.

HClausing committed:
Fixed double css class declaration on example
Added some tests

henon commented:
The blue slide background is pretty dark and would look much better with white text, arrows and delimiters. But I guess it will not be easy to have a different arrow color etc. for only one of the slides, right?
![image](https://user-images.githubusercontent.com/44090/115967263-ee01ad80-a531-11eb-8685-8949e189bedb.png)

henon commented:
Test coverage is reasonable I think:
![image](https://user-images.githubusercontent.com/44090/115967423-d8d94e80-a532-11eb-9ed2-4f8e9b4f9600.png)
![image](https://user-images.githubusercontent.com/44090/115967499-32417d80-a533-11eb-9021-a3f2c99d4741.png)

Just one thing. You introduced the MudBlazor.Base namespace. All other base classes are living in MudBlazor, so I think we either remove that namespace. We can later decide to put all base classes in a base namespace.

HClausing commented:
> Test coverage is reasonable I think:
> ![image](https://user-images.githubusercontent.com/44090/115967423-d8d94e80-a532-11eb-9ed2-4f8e9b4f9600.png)
> ![image](https://user-images.githubusercontent.com/44090/115967499-32417d80-a533-11eb-9021-a3f2c99d4741.png)
> 
> Just one thing. You introduced the MudBlazor.Base namespace. All other base classes are living in MudBlazor, so I think we better remove that namespace. We can later decide to put all base classes in a base namespace.
> @mikes-gh what do you think?
> @Garderoben please have a look. The transitions are just beautiful. Are you content with the docs page as it is now?

Nice, thanks for running the test coverage, I don't have the tools, and that open source that I told you requires some configs on the project files.

I guess that I can get some better results on theses tests. Please let me try.

About the namespace, I'm sorry, it's my mistake. I can fix on next commit until the final decision.

HClausing commented:
> The blue slide background is pretty dark and would look much better with white text, arrows and delimiters. But I guess it will not be easy to have a different arrow color etc. for only one of the slides, right?
> ![image](https://user-images.githubusercontent.com/44090/115967263-ee01ad80-a531-11eb-8685-8949e189bedb.png)

Sure, we have these parameters:
LeftArrowTemplate ;
DelimiterTemplate ;
RightArrowTemplate .

By default, they are all null, so the component uses these default buttons with theme text color.

You gives me a good ideia: I will create a new example on Docs combining if () and SelectedIndex on these parameters.

HClausing committed:
- Implemented theme color compatibility for navigation controls;
 - Added example using DelimiterTemplate, LeftArrowTemplate, and LeftArrowTemplate;
 - Added more tests.

HClausing commented:
Based on last talk, I've made some changes:

 - Added color themes for navigation controls:

![theme_compatible](https://user-images.githubusercontent.com/15158923/116020570-f387f200-a61c-11eb-96a3-dd0faf73bec8.gif)

 - Added an Example demonstrating how to work with control's templates (delimiters and arrows):

![templates](https://user-images.githubusercontent.com/15158923/116020625-10bcc080-a61d-11eb-99f1-528984cfb021.gif)

HClausing committed:
Changes on CarouselTest_AutoCycle to ensure success on pipeline tests.

mikes-gh commented:
@HClausing FYI The code coverage report is included in every build.  Got show all checks and drill down to the code coverage.

<img width="488" alt="Screenshot 2021-04-26 at 22 49 14" src="https://user-images.githubusercontent.com/16208742/116155321-dc5f0800-a6e1-11eb-9565-e55ccc5012e9.png">

JonBunator commented:
Looks really good! But I found one small issue in RTL mode:
The icons need to be flipped 180 degrees. See picture below

![carousel](https://user-images.githubusercontent.com/62108893/117423261-d1e21100-af20-11eb-857a-aa80c4592998.png)

You can try it out by clicking on the icon on the top right (appbar)
![grafik](https://user-images.githubusercontent.com/62108893/117423438-0229af80-af21-11eb-8fa9-858937db64fb.png)

JonBunator commented on src/MudBlazor.Docs/Pages/Components/Carousel/Examples/CarouselExample.razor:
```diff
@@ -0,0 +1,33 @@
+        </div>
+    </MudCarouselItem>
+</MudCarousel>
+<MudSelect @bind-Value="transition" Label="Transition" Class="ma-2">
```
You could add
```csharp
<MudSelectItem Value="@Transition.None">None</MudSelectItem>
```
as a option

HClausing replied:
Thanks for let me know @JonBunator .

One doublt: Should the first page stay on right too when RTL is active?

HClausing replied:
I agree with you, but I guess that the flippers order must be inverted too, because the last flipper must be the first page, and not the last one, because the .Next() should cause a visually move to que left flipper, and not to the right one.

HClausing replied:
Exaclty, and the Radio button from the right side must be the first page of the collection index

HClausing replied:
Thanks for help on the RTL tests @JonBunator. It's supported now.

JonBunator commented on src/MudBlazor.Docs/Pages/Components/Carousel/Examples/CarouselExample.razor:
```diff
@@ -0,0 +1,33 @@
+﻿@namespace MudBlazor.Docs.Examples
```
I found a strange behaviour in this example. When the transition ist set to `Transition.None` the third item changes its color.
![example3](https://user-images.githubusercontent.com/62108893/117449588-73c52600-af40-11eb-89f9-a586a276cdb4.gif)

Third Item before:
![example1](https://user-images.githubusercontent.com/62108893/117449666-85a6c900-af40-11eb-85ad-5ef64ef67b56.png)

Third item after:
![example2](https://user-images.githubusercontent.com/62108893/117449689-8b9caa00-af40-11eb-85b5-acf9734fca7f.png)

HClausing replied:
I'll take a look. I forgot to test Transition.None. Thanks.

HClausing replied:
Fixed! Thanks!

HClausing committed:
Fixes and improvements

 - Renamed LeftArrowTemplate and RightArrowTemplate to PreviousButtonTemplate and NextButtonTemplate;
 - Added support for RTL;
 - Fixed rendering issues when Transition = None and Items with inherited background color;

So many thanks for all that involved with this PR, reviewing and sending ideias.

henon commented:
Thanks for your awesome work Henrique! Garderoben decided to fix the style tags later when he got time.
